### ISSUE TITLE
[Python] Early Stopping with Best Model Checkpoint

### ISSUE THREAD

**Issue by Aamod007** [etsi-ai/etna#39]
## Description

Training currently runs for a fixed number of epochs regardless of model performance. Implementing early stopping will automatically halt training when the model stops improving, preventing overfitting and saving computation time.

## Current Behavior

```python
model.train(epochs=100, lr=0.01)
# Runs all 100 epochs even if loss plateaus at epoch 20
```

## Expected Behavior

```python
model.train(
    epochs=100, 
    lr=0.01,
    early_stopping=True,
    patience=10,           # Stop if no improvement for 10 epochs
    restore_best=True      # Restore weights from best epoch
)
# Stops at epoch 30 if loss stopped improving at epoch 20
# Model weights are from epoch 20 (best), not epoch 30 (last)
```

## Implementation Details

1. Add `early_stopping`, `patience`, and `restore_best` parameters to `Model.train()`
2. Track validation/training loss history during training
3. Implement patience counter - stop training if loss doesn't improve for `patience` consecutive epochs
4. Cache model weights at best epoch and restore them when training stops
5. Print informative message when early stopping triggers

## Acceptance Criteria

- [ ] Training stops automatically when loss stops improving
- [ ] Best model weights are restored (not the last epoch weights)
- [ ] Works for both classification and regression tasks
- [ ] Backward compatible - default behavior unchanged if `early_stopping=False`

## Related Issues

- #22 (Validation Split) - Early stopping works best with validation loss tracking

**Comment by reddyeswaranush** [comment]
Hi @Aamod007 , I’m interested in working on this issue as part of SWoC’26. I plan to extend Model.train() with optional parameters for early stopping (early_stopping, patience, restore_best) while keeping the default behavior backward compatible. The implementation will track loss per epoch, maintain a best-loss checkpoint with a patience counter, stop training when the loss stops improving for a defined number of epochs, and restore the best model weights if enabled. This approach will work consistently for both classification and regression tasks and remain compatible with the existing MLflow tracking setup. Please let me know if I can proceed and assign this issue to me. Thanks!

**Comment by Satyamgupta2365** [comment]
@Gyanprakash136  join the discord and tell your plan

**Comment by Gyanprakash136** [comment]
Sure

**Comment by Madhavi1108** [comment]
Hi,
I would like to work on this issue
